File: modulo08-arquitetura-de-sistemas-com-ia/modulo-05-arquitetura-enterprise/trialforge-model-tiering-prototype-go/tiering/vector_math.go
package tiering

import "math"

// SimilaridadeCosseno é lógica pura, testável sem rede.
func SimilaridadeCosseno(a, b []float64) float64 {
	var produto, normaA, normaB float64
	for i := range a {
		produto += a[i] * b[i]
		normaA += a[i] * a[i]
		normaB += b[i] * b[i]
	}
	return produto / (math.Sqrt(normaA) * math.Sqrt(normaB))
}

File: modulo08-arquitetura-de-sistemas-com-ia/modulo-05-arquitetura-enterprise/trialforge-model-tiering-prototype-java/src/main/java/com/trialforge/tiering/VectorMath.java
package com.trialforge.tiering;

/** Matemática de vetores usada pelo RAG e pela confiança de resposta — lógica pura, sem rede. */
public final class VectorMath {

    private VectorMath() {
    }

    public static double similaridadeCosseno(double[] a, double[] b) {
        double produto = 0, normaA = 0, normaB = 0;
        for (int i = 0; i < a.length; i++) {
            produto += a[i] * b[i];
            normaA += a[i] * a[i];
            normaB += b[i] * b[i];
        }
        return produto / (Math.sqrt(normaA) * Math.sqrt(normaB));
    }
}
